pkg/retriever/util/util.go

```diff
@@ -0,0 +1,119 @@
+package util
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/filecoin-project/lassie/pkg/types"
+	"go.uber.org/multierr"
+)
+
+func WithCandidates(retrievalCandidateFinder types.RetrievalCandidateFinder, childRetriever types.CandidateRetriever) types.Retriever {
+	return func(ctx context.Context, request types.RetrievalRequest, events func(types.RetrievalEvent)) (*types.RetrievalStats, error) {
+		candidates, err := retrievalCandidateFinder(ctx, request, events)
+		if err != nil {
+			return nil, err
+		}
+		return childRetriever(ctx, request, candidates, events)
+	}
+}
+
+func WithFoundCandidates(candidates []types.RetrievalCandidate, candidateRetriever types.CandidateRetriever) types.Retriever {
+	return func(ctx context.Context, request types.RetrievalRequest, events func(types.RetrievalEvent)) (*types.RetrievalStats, error) {
+		return candidateRetriever(ctx, request, candidates, events)
+	}
+}
+
+func collectResults(ctx context.Context, resultChan <-chan types.RetrievalResult, totalRetrievers int) (*types.RetrievalStats, error) {
+	var totalErr error
+	finishedCount := 0
+	for {
+		select {
+		case result := <-resultChan:
+			if result.Err != nil {
+				totalErr = multierr.Append(totalErr, result.Err)
+			}
+			if result.Stats != nil {
+				return result.Stats, nil
+			}
+			// have we got all responses but no success?
+			finishedCount++
+			if finishedCount >= totalRetrievers {
+				return nil, totalErr
+			}
+		case <-ctx.Done():
+			return nil, context.Canceled
+		}
+	}
+}
+
+// TODO: seems like we should be able to adapt the GraphsyncRetriever to use this
+
+func RaceRetrievers(retrievers []types.Retriever) types.Retriever {
```

Comment:
I'm personally not a big fan of util files since they don't really help describe what they contain as they become big catch alls. These Retrievers feel like individually standing things. Could we not throw them into their own files?